pkg/provider/azure/keyvault/keyvault.go

```diff
@@ -233,55 +234,94 @@ func (a *Azure) GetAllSecrets(ctx context.Context, ref esv1beta1.ExternalSecretF
 
-	if ref.Version != "" {
-		version = ref.Version
-	}
-
 	switch objectType {
 	case defaultObjType:
 		// returns a SecretBundle with the secret value
 		// https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#SecretBundle
-		secretResp, err := a.baseClient.GetSecret(context.Background(), *a.provider.VaultURL, secretName, version)
+		secretResp, err := a.baseClient.GetSecret(context.Background(), *a.provider.VaultURL, secretName, ref.Version)
 		if err != nil {
 			return nil, err
 		}
-		if ref.Property == "" {
-			return []byte(*secretResp.Value), nil
-		}
-		res := gjson.Get(*secretResp.Value, ref.Property)
-		if !res.Exists() {
-			return nil, fmt.Errorf(errPropNotExist, ref.Property, ref.Key)
+		if ref.MetadataPolicy == esv1beta1.ExternalSecretMetadataPolicyFetch {
+			return getSecretTag(secretResp.Tags, ref.Property)
 		}
-		return []byte(res.String()), err
+		return getProperty(*secretResp.Value, ref.Property, ref.Key, err)
 	case objectTypeCert:
 		// returns a CertBundle. We return CER contents of x509 certificate
 		// see: https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#CertificateBundle
-		secretResp, err := a.baseClient.GetCertificate(context.Background(), *a.provider.VaultURL, secretName, version)
+		certResp, err := a.baseClient.GetCertificate(context.Background(), *a.provider.VaultURL, secretName, ref.Version)
 		if err != nil {
 			return nil, err
 		}
-		return *secretResp.Cer, nil
+		if ref.MetadataPolicy == esv1beta1.ExternalSecretMetadataPolicyFetch {
+			return getSecretTag(certResp.Tags, ref.Property)
+		}
+		return *certResp.Cer, nil
 	case objectTypeKey:
 		// returns a KeyBundle that contains a jwk
 		// azure kv returns only public keys
 		// see: https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#KeyBundle
-		keyResp, err := a.baseClient.GetKey(context.Background(), *a.provider.VaultURL, secretName, version)
+		keyResp, err := a.baseClient.GetKey(context.Background(), *a.provider.VaultURL, secretName, ref.Version)
 		if err != nil {
 			return nil, err
 		}
+		if ref.MetadataPolicy == esv1beta1.ExternalSecretMetadataPolicyFetch {
+			return getSecretTag(keyResp.Tags, ref.Property)
+		}
 		return json.Marshal(keyResp.Key)
+
 	}
 
 	return nil, fmt.Errorf(errUnknownObjectType, secretName)
 }
 
+func (a *Azure) getSecretTags(ref esv1beta1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	_, secretName := getObjType(ref)
+
+	// returns a SecretBundle with the secret value
+	// https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#SecretBundle
+	secretResp, err := a.baseClient.GetSecret(context.Background(), *a.provider.VaultURL, secretName, ref.Version)
+	if err != nil {
+		return nil, err
+	}
+	secretTagsData := make(map[string][]byte)
+	for k, v := range secretResp.Tags {
+		secretTagsData[k] = []byte(*v)
+	}
+	return secretTagsData, nil
+}
```

Comment:
Can we put a JSON inside a custom metadata for Azure KeyVault? If yes, I think we need to add the possibility to load it here (similar with `getProperty` method)

Reply:
My idea here was the following (this is the same behavior for Secrets in Azure KeyVault):

If I have the following metadata called meta:
```
"foo": "bar"
"custom": {"foo":"yada"}
```
and if I have the following dataFrom:

```
dataFrom:
- extract:
    metadataPolicy: Fetch
    key: meta
    property: custom
```
I expect to see a kubernetes secret of the fashion:
```
foo: yada
```
currently I believe dataFrom for tags would output the following (no matter what is within property):
```
foo: bar
custom: {"foo":"yada"}
```

Reply:
but if you only want one tag you can use:

```
data:
- secretKey: something
    metadataPolicy: Fetch
    key: meta
    property: custom
```

and you'll get the foo: yada

do we also want the same behavior in dataFrom?

Reply:
I think the behavior is going to be different though:

```
# data example
data:
- secretKey: something
    metadataPolicy: Fetch
    key: meta
    property: custom
 # Generates
 something: {"foo":"yada"}
# dataFrom example
dataFrom:
- extract:
-  metadataPolicy: Fetch
    key: meta
    property: custom
 # Generates
 foo: yada
```
dataFrom is unpacking the JSON, while data is not.

Reply:
Hello! To me it may makes sense, since this way we would be providing users different ways to retrieve the secrets metadata. Do you think that's better being consistent in both situation?

Reply:
The thing is, the current behavior doesn't unpack JSON secrets in any case, with data

```
data:
- secretKey: something
  remoteRef:
    metadataPolicy: Fetch
    key: meta
    property: custom
```

Generates
```
something: {"foo":"yada"}
```
right, but if you use dataFrom

```
dataFrom:
- extract:
-  metadataPolicy: Fetch
    key: meta
    property: custom
```
You'll get 

```
foo: bar
custom: {"foo":"yada"}
```

We can add the feature that allows dataFrom to only get custom tag as specified in property, but is going to be in the same format. Is it preferred to have this option twice?

Reply:
Hi @dreadful-dragon ! My point is a little bit different. If we specify `MetadataPolicy: Fetch` or `MetadataPolicy: None`, the behavior of `data` and `dataForm.extract` should not change. With the current implementation, it is changing. This is happening because `dataFrom.extract` when `MetadataPolicy:None` is fetching a secret Key, and then decoding the JSON from that key. For the `MetadataPolicy: Fetch`, we are ignoring `property` whatsoever. What I would expect is: If a given Secret has the same metadata information as the data, so a secret like:

```
data: {"foo":{"nested":"bar"}}
metadata: "{foo":{"nested":"bar"}}
```

The outcomes of data /dataFrom combinations using `MetadataPolicy:None` and `MetadataPolicy:Fetch` should be the same.

Reply:
Hello again @gusfcarvalho, maybe I'm wrong but in your last example when I try with data

```
data:
- secretKey: something
  remoteRef:
    key: data
```
Generates
```
{"foo":{"nested":"bar"}}
```
and when I try to retrieve the tags 

```
data:
- secretKey: something
  remoteRef:
    metadataPolicy: Fetch
    key: data
```
Generates
```
{"metadata":"{\"foo\":{\"nested\":\"bar\"}}}
```
because when a tag is not specified you get all of them in JSON format. 
Although when it is specified:

```
data:
- secretKey: something
  remoteRef:
    metadataPolicy: Fetch
    property: metadata
    key: data
```
Generates
```
{"foo":{"nested":"bar"}}
```

Also, when I'm using datafrom 

```
dataFrom:
- extract:
    key: data
```
Generates
```
{"foo":{"nested":"bar"}}
```
and try retrieving the tags
```
dataFrom:
- extract:
    key: data
    metadataPolicy: Fetch
```
Also generates
```
{"foo":{"nested":"bar"}}
```

In any case datafrom is not unpacking, not when synchronizing the secret value or the tags, so I would say is consistent.

But again, I might be doing something wrong and if you prefer I could skip the metadataPolicy in datafrom for now, so data could be already in use.

Reply:
Hello @gusfcarvalho, I think I finally got your point.
Now datafrom.extract will decode the JSON tags.
Let me know if it's what you had in mind.

I have also added the possibility to use metadataPolicy with datafrom.find so all the resources plus all the tags from every resource are retrieved.